File: core-api/src/main/java/com/emarsys/core/api/result/Try.java
package com.emarsys.core.api.result;

import androidx.annotation.Nullable;

public class Try<T> {

    public static <T> Try success(T result) {
        return new Try<>(result, null);
    }

    public static <T> Try failure(Exception errorCause) {
        return new Try<>((T) null, errorCause);
    }

    private final T result;
    private final Throwable errorCause;

    public Try(T result, Throwable errorCause) {
        this.result = result;
        this.errorCause = errorCause;
    }

    @Nullable
    public T getResult() {
        return result;
    }

    @Nullable
    public Throwable getErrorCause() {
        return errorCause;
    }
}

File: core-api/src/main/java/com/emarsys/core/api/result/Try.kt
package com.emarsys.core.api.result

class Try<T>(val result: T?, val errorCause: Throwable?) {

    companion object {
        @JvmStatic
        fun <T> success(result: T): Try<T> {
            return Try(result, null)
        }

        @JvmStatic
        fun <T> failure(errorCause: Exception?): Try<T> {
            return Try(null as T?, errorCause)
        }
    }
}